File: translate/TranslateExampleObjC/ViewController.m
//
// Copyright (c) 2019 Google Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "ViewController.h"
@import Firebase;

NS_ASSUME_NONNULL_BEGIN

@interface ViewController ()

@property(weak, nonatomic) IBOutlet UITextView *inputTextView;
@property(weak, nonatomic) IBOutlet UITextView *outputTextView;
@property(weak, nonatomic) IBOutlet UITextView *statusTextView;
@property(weak, nonatomic) IBOutlet UIPickerView *inputPicker;
@property(weak, nonatomic) IBOutlet UIPickerView *outputPicker;
@property(weak, nonatomic) IBOutlet UIButton *sourceDownloadDeleteButton;
@property(weak, nonatomic) IBOutlet UIButton *targetDownloadDeleteButton;

@property(nonatomic, strong) FIRTranslator *translator;
@property(nonatomic, strong) NSArray<NSNumber *> *allLanguages;

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  self.allLanguages = FIRTranslateAllLanguages().allObjects;
  self.inputPicker.dataSource = self;
  self.outputPicker.dataSource = self;
  [self.inputPicker selectRow:[self.allLanguages indexOfObject:@(FIRTranslateLanguageEN)]
                  inComponent:0
                     animated:NO];
  [self.outputPicker selectRow:[self.allLanguages indexOfObject:@(FIRTranslateLanguageES)]
                   inComponent:0
                      animated:NO];
  self.inputPicker.delegate = self;
  self.outputPicker.delegate = self;
  self.inputTextView.delegate = self;
  self.inputTextView.returnKeyType = UIReturnKeyDone;

  [self pickerView:self.inputPicker didSelectRow:0 inComponent:0];
  [self updateDownloadDeleteButtonLabels];

  [NSNotificationCenter.defaultCenter
      addObserver:self
         selector:@selector(modelDownloadDidCompleteWithNotification:)
             name:FIRModelDownloadDidSucceedNotification
           object:nil];
  [NSNotificationCenter.defaultCenter
      addObserver:self
         selector:@selector(modelDownloadDidCompleteWithNotification:)
             name:FIRModelDownloadDidFailNotification
           object:nil];
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
  return 1;
}

- (nullable NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
  return FIRTranslateLanguageCodeForLanguage(self.allLanguages[row].integerValue);
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
  return self.allLanguages.count;
}

- (void)textViewDidChange:(UITextView *)textView {
  [self translate];
}

- (BOOL)textView:(UITextView *)textView
    shouldChangeTextInRange:(NSRange)range
            replacementText:(NSString *)text {
  // Hide the keyboard when "Done" is pressed.
  // See: https://stackoverflow.com/questions/26600359/dismiss-keyboard-with-a-uitextview
  if ([text isEqualToString:@"\n"]) {
    [textView resignFirstResponder];
    return NO;
  }
  return YES;
}

- (IBAction)didTapSwap {
  NSInteger inputSelectedRow = [self.inputPicker selectedRowInComponent:0];
  [self.inputPicker selectRow:[self.outputPicker selectedRowInComponent:0] inComponent:0 animated:NO];
  [self.outputPicker selectRow:inputSelectedRow inComponent:0 animated:NO];
  self.inputTextView.text = self.outputTextView.text;
  [self pickerView:self.inputPicker didSelectRow:0 inComponent:0];
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
  FIRTranslatorOptions *options =
  [[FIRTranslatorOptions alloc] initWithSourceLanguage:self.allLanguages[[self.inputPicker selectedRowInComponent:0]].integerValue
                                        targetLanguage:self.allLanguages[[self.outputPicker selectedRowInComponent:0]].integerValue];
  self.translator = [[FIRNaturalLanguage naturalLanguage] translatorWithOptions:options];
  [self translate];
}

- (void)translate {
  FIRTranslator *translatorForDownload = self.translator;
  [self.translator downloadModelIfNeededWithCompletion:^(NSError *_Nullable error) {
    if (error != nil) {
      self.outputTextView.text =
          [NSString stringWithFormat:@"Failed to ensure model downloaded with error %@",
                                     error.localizedDescription];
      return;
    }
    [self updateDownloadDeleteButtonLabels];
    NSString *text = self.inputTextView.text;
    if (text == nil) {
      text = @"";
    }
    self.outputTextView.text = @"";
    if (translatorForDownload != self.translator) {
      return;
    }
    [self.translator translateText:text
                        completion:^(NSString *_Nullable result, NSError *_Nullable error) {
                          if (error != nil) {
                            self.outputTextView.text = [NSString
                                stringWithFormat:@"Failed to ensure model downloaded with error %@",
                                                 error.localizedDescription];
                            return;
                          }
                          if (translatorForDownload != self.translator) {
                            return;
                          }
                          self.outputTextView.text = result;
                        }];
  }];
}

- (void)handleDownloadDeleteWithPicker:(UIPickerView *)picker button:(UIButton *)button {
  FIRTranslateLanguage language =
      (FIRTranslateLanguage)self.allLanguages[[picker selectedRowInComponent:0]].intValue;
  NSString *languageCode = FIRTranslateLanguageCodeForLanguage(language);

  [button setTitle:@"Working..." forState:UIControlStateNormal];
  FIRTranslateRemoteModel *model = [self modelForLanguage:language];
  FIRModelManager *modelManager = [FIRModelManager modelManager];

  if ([modelManager isModelDownloaded:model]) {
    self.statusTextView.text = [NSString stringWithFormat:@"Deleting %@", languageCode];
    [modelManager deleteDownloadedModel:model
                             completion:^(NSError *_Nullable error) {
                               [self updateDownloadDeleteButtonLabels];
                               self.statusTextView.text =
                                   [NSString stringWithFormat:@"Deleted %@", languageCode];
                             }];
  } else {
    self.statusTextView.text = [NSString stringWithFormat:@"Downloading %@", languageCode];
    FIRModelDownloadConditions *conditions =
        [[FIRModelDownloadConditions alloc] initWithAllowsCellularAccess:YES
                                             allowsBackgroundDownloading:YES];
    [modelManager downloadModel:model conditions:conditions];
  }
}

- (void)updateDownloadDeleteButtonLabels {
  FIRTranslateLanguage inputLanguage =
      (FIRTranslateLanguage)self.allLanguages[[self.inputPicker selectedRowInComponent:0]].intValue;
  FIRTranslateLanguage outputLanguage =
      (FIRTranslateLanguage)self.allLanguages[[self.outputPicker selectedRowInComponent:0]]
          .intValue;

  if ([self isLanguageDownloaded:inputLanguage]) {
    [self.sourceDownloadDeleteButton setTitle:@"Delete Model" forState:UIControlStateNormal];
  } else {
    [self.sourceDownloadDeleteButton setTitle:@"Download Model" forState:UIControlStateNormal];
  }
  if ([self isLanguageDownloaded:outputLanguage]) {
    [self.targetDownloadDeleteButton setTitle:@"Delete Model" forState:UIControlStateNormal];
  } else {
    [self.targetDownloadDeleteButton setTitle:@"Download Model" forState:UIControlStateNormal];
  }
}

- (BOOL)isLanguageDownloaded:(FIRTranslateLanguage)language {
  FIRTranslateRemoteModel *model = [self modelForLanguage:language];
  FIRModelManager *modelManager = [FIRModelManager modelManager];
  return [modelManager isModelDownloaded:model];
}

- (IBAction)listDownloadedModels {
  FIRModelManager *modelManager = [FIRModelManager modelManager];
  NSMutableString *listOfLanguages = [NSMutableString string];
  for (FIRTranslateRemoteModel *model in modelManager.downloadedTranslateModels) {
    if (listOfLanguages.length > 0) {
      [listOfLanguages appendString:@", "];
    }
    [listOfLanguages appendString:FIRTranslateLanguageCodeForLanguage(model.language)];
  }
  self.statusTextView.text = [NSString stringWithFormat:@"Downloaded models: %@", listOfLanguages];
}

- (IBAction)didTapDownloadDeleteSourceLanguage {
  [self handleDownloadDeleteWithPicker:self.inputPicker button:self.sourceDownloadDeleteButton];
}

- (IBAction)didTapDownloadDeleteTargetLanguage {
  [self handleDownloadDeleteWithPicker:self.outputPicker button:self.targetDownloadDeleteButton];
}

- (FIRTranslateRemoteModel *)modelForLanguage:(FIRTranslateLanguage)language {
  return [FIRTranslateRemoteModel translateRemoteModelWithLanguage:language];
}

- (void)modelDownloadDidCompleteWithNotification:(NSNotification *)notification {
  FIRTranslateRemoteModel *model = notification.userInfo[FIRModelDownloadUserInfoKeyRemoteModel];
  if (![model isKindOfClass:FIRTranslateRemoteModel.class]) {
    return;
  }
  NSString *languageCode = FIRTranslateLanguageCodeForLanguage(model.language);

  dispatch_async(dispatch_get_main_queue(), ^{
    if (notification.name == FIRModelDownloadDidSucceedNotification) {
      self.statusTextView.text =
          [NSString stringWithFormat:@"Download succeeded for %@", languageCode];
    } else {
      self.statusTextView.text = [NSString stringWithFormat:@"Download failed for%@", languageCode];
    }
    [self updateDownloadDeleteButtonLabels];
  });
}

@end

NS_ASSUME_NONNULL_END

File: translate/TranslateExample/ViewController.swift
//
// Copyright (c) 2019 Google Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import UIKit

import Firebase

@objc(ViewController)
class ViewController: UIViewController, UITextViewDelegate, UIPickerViewDataSource, UIPickerViewDelegate {

  @IBOutlet var inputTextView: UITextView!
  @IBOutlet var outputTextView: UITextView!
  @IBOutlet var statusTextView: UITextView!
  @IBOutlet var inputPicker: UIPickerView!
  @IBOutlet var outputPicker: UIPickerView!
  @IBOutlet var sourceDownloadDeleteButton: UIButton!
  @IBOutlet var targetDownloadDeleteButton: UIButton!

  var translator: Translator!
  lazy var allLanguages = TranslateLanguage.allLanguages().compactMap {
    TranslateLanguage(rawValue: $0.uintValue)
  }

  override func viewDidLoad() {
    inputPicker.dataSource = self
    outputPicker.dataSource = self
    inputPicker.selectRow(allLanguages.firstIndex(of: TranslateLanguage.en) ?? 0, inComponent: 0, animated: false)
    outputPicker.selectRow(allLanguages.firstIndex(of: TranslateLanguage.es) ?? 0, inComponent: 0, animated: false)
    inputPicker.delegate = self
    outputPicker.delegate = self
    inputTextView.delegate = self
    inputTextView.returnKeyType = .done
    pickerView(inputPicker, didSelectRow: 0, inComponent: 0)
    setDownloadDeleteButtonLabels()

    NotificationCenter.default.addObserver(self, selector:#selector(remoteModelDownloadDidComplete(notification:)), name:.firebaseMLModelDownloadDidSucceed, object:nil)
    NotificationCenter.default.addObserver(self, selector:#selector(remoteModelDownloadDidComplete(notification:)), name:.firebaseMLModelDownloadDidFail, object:nil)
  }

  func numberOfComponents(in pickerView: UIPickerView) -> Int {
    return 1
  }

  func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
    return allLanguages[row].toLanguageCode()
  }

  func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
    return allLanguages.count
  }

  func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange,
                replacementText text: String) -> Bool {
    // Hide the keyboard when "Done" is pressed.
    // See: https://stackoverflow.com/questions/26600359/dismiss-keyboard-with-a-uitextview
    if (text == "\n") {
      textView.resignFirstResponder()
      return false
    }
    return true
  }

  func textViewDidChange(_ textView: UITextView) {
    translate()
  }

  @IBAction func didTapSwap() {
    let inputSelectedRow = inputPicker.selectedRow(inComponent: 0)
    inputPicker.selectRow(outputPicker.selectedRow(inComponent: 0), inComponent: 0, animated: false)
    outputPicker.selectRow(inputSelectedRow, inComponent: 0, animated: false)
    inputTextView.text = outputTextView.text
    pickerView(inputPicker, didSelectRow: 0, inComponent: 0)
  }

  func model(forLanguage: TranslateLanguage) -> TranslateRemoteModel {
    return TranslateRemoteModel.translateRemoteModel(language: forLanguage)
  }

  func isLanguageDownloaded(_ language: TranslateLanguage) -> Bool {
    let model = self.model(forLanguage: language)
    let modelManager = ModelManager.modelManager()
    return modelManager.isModelDownloaded(model)
  }

  func handleDownloadDelete(picker: UIPickerView, button: UIButton) {
    let language = allLanguages[picker.selectedRow(inComponent: 0)]
    button.setTitle("working...", for: .normal)
    let model = self.model(forLanguage: language)
    let modelManager = ModelManager.modelManager()
    if modelManager.isModelDownloaded(model) {
      self.statusTextView.text = "Deleting " + language.toLanguageCode()
      modelManager.deleteDownloadedModel(model) { error in
        self.statusTextView.text = "Deleted " + language.toLanguageCode()
        self.setDownloadDeleteButtonLabels()
      }
    } else {
      self.statusTextView.text = "Downloading " + language.toLanguageCode()
      let conditions = ModelDownloadConditions(
        allowsCellularAccess: true,
        allowsBackgroundDownloading: true
      )
      modelManager.download(model, conditions:conditions)
    }
  }

  @IBAction func didTapDownloadDeleteSourceLanguage() {
    self.handleDownloadDelete(picker: inputPicker, button: self.sourceDownloadDeleteButton)
  }

  @IBAction func didTapDownloadDeleteTargetLanguage() {
    self.handleDownloadDelete(picker: outputPicker, button: self.targetDownloadDeleteButton)
  }

  @IBAction func listDownloadedModels() {
    let msg = "Downloaded models:" + ModelManager.modelManager()
      .downloadedTranslateModels
      .map { model in model.language.toLanguageCode() }
      .joined(separator: ", ");
    self.statusTextView.text = msg
  }

  @objc
  func remoteModelDownloadDidComplete(notification: NSNotification) {
    let userInfo = notification.userInfo!
    guard let remoteModel =
      userInfo[ModelDownloadUserInfoKey.remoteModel.rawValue] as? TranslateRemoteModel else {
        return
    }
    DispatchQueue.main.async {
      if notification.name == .firebaseMLModelDownloadDidSucceed {
        self.statusTextView.text = "Download succeeded for " + remoteModel.language.toLanguageCode()
      } else {
        self.statusTextView.text = "Download failed for " + remoteModel.language.toLanguageCode()
      }
      self.setDownloadDeleteButtonLabels()
    }
  }

  func setDownloadDeleteButtonLabels() {
    let inputLanguage = allLanguages[inputPicker.selectedRow(inComponent: 0)]
    let outputLanguage = allLanguages[outputPicker.selectedRow(inComponent: 0)]
    if self.isLanguageDownloaded(inputLanguage) {
      self.sourceDownloadDeleteButton.setTitle("Delete model", for: .normal)
    } else {
      self.sourceDownloadDeleteButton.setTitle("Download model", for: .normal)
    }
    if self.isLanguageDownloaded(outputLanguage) {
      self.targetDownloadDeleteButton.setTitle("Delete model", for: .normal)
    } else {
      self.targetDownloadDeleteButton.setTitle("Download model", for: .normal)
    }
  }

  func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
    let inputLanguage = allLanguages[inputPicker.selectedRow(inComponent: 0)]
    let outputLanguage = allLanguages[outputPicker.selectedRow(inComponent: 0)]
    self.setDownloadDeleteButtonLabels()
    let options = TranslatorOptions(sourceLanguage: inputLanguage, targetLanguage: outputLanguage)
    translator = NaturalLanguage.naturalLanguage().translator(options: options)
    translate()
  }

  func translate() {
    let translatorForDownloading = self.translator!

    translatorForDownloading.downloadModelIfNeeded { error in
      guard error == nil else {
        self.outputTextView.text = "Failed to ensure model downloaded with error \(error!)"
        return
      }
      self.setDownloadDeleteButtonLabels()
      if translatorForDownloading == self.translator {
        translatorForDownloading.translate(self.inputTextView.text ?? "") { result, error in
          guard error == nil else {
            self.outputTextView.text = "Failed with error \(error!)"
            return
          }
          if translatorForDownloading == self.translator {
            self.outputTextView.text = result
          }
        }
      }
    }
  }
}
